controllers/cloud_config_sync_controller.go

```diff
@@ -0,0 +1,196 @@
+		}
+	}
+	err = r.Get(ctx, sourceConfigMapKey, sourceCM)
+	if err != nil {
+		klog.Errorf("unable to get source cloud-config: %v", err)
+		return ctrl.Result{}, err
+	}
+	sourceCM, err = r.prepareSourceConfigMap(sourceCM)
+	if err != nil {
+		klog.Errorf("unable to check and prepare source cloud-config: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	targetConfigMapKey := client.ObjectKey{
+		Namespace: r.TargetNamespace,
+		Name:      cloudConfigMapName,
+	}
+	targetCM := &corev1.ConfigMap{}
+	err = r.Get(ctx, targetConfigMapKey, targetCM)
+	if err != nil && errors.IsNotFound(err) {
+		klog.Infof("target cloud-config not found in namespace %s, syncing", r.TargetNamespace)
+		err = r.syncCloudConfigData(ctx, sourceCM, targetCM)
+		if err != nil {
+			klog.Errorf("unable to sync cloud-config: %v", err)
+			return ctrl.Result{}, err
+		}
+	}
+	if err != nil {
+		klog.Errorf("unable to obtain target cloud-config: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	if r.isCloudConfigEqual(sourceCM, targetCM) {
+		klog.Infof("source and target cloud-configs content are equal, no sync needed")
+		return ctrl.Result{}, nil
+	} else {
+		err = r.syncCloudConfigData(ctx, sourceCM, targetCM)
+		if err != nil {
+			klog.Errorf("unable to sync cloud-config: %v", err)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+func (r *CloudConfigReconciler) prepareSourceConfigMap(source *corev1.ConfigMap) (*corev1.ConfigMap, error) {
+	// Keys might be different between openshift-config/cloud-config and openshift-config-managed/kube-cloud-config
+	// Always use "cloud.conf" which is default one across openshift
+	cloudConfCm := source.DeepCopy()
+	if _, ok := cloudConfCm.Data[defaultConfigKey]; ok {
+		return cloudConfCm, nil
+	}
+
+	infraConfigKey := r.InfrastructureResource.Spec.CloudConfig.Key
+
+	if val, ok := cloudConfCm.Data[infraConfigKey]; ok {
+		cloudConfCm.Data[defaultConfigKey] = val
+		delete(cloudConfCm.Data, infraConfigKey)
+		return cloudConfCm, nil
+	}
+
+	return nil, fmt.Errorf("default key %s does not found in source configmap %s", defaultConfigKey, client.ObjectKeyFromObject(source))
+}
+
+func (r *CloudConfigReconciler) isCloudConfigEqual(source *corev1.ConfigMap, target *corev1.ConfigMap) bool {
+	return source.Immutable == target.Immutable &&
+		reflect.DeepEqual(source.Data, target.Data) && reflect.DeepEqual(source.BinaryData, target.BinaryData)
+}
+
+func (r *CloudConfigReconciler) syncCloudConfigData(ctx context.Context, source *corev1.ConfigMap, target *corev1.ConfigMap) error {
+
+	target.SetName(cloudConfigMapName)
+	target.SetNamespace(r.TargetNamespace)
+	target.Data = source.Data
+	target.BinaryData = source.BinaryData
+	target.Immutable = source.Immutable
+
+	// check if target config exists, crate if not
+	err := r.Get(ctx, client.ObjectKeyFromObject(target), &corev1.ConfigMap{})
+
+	if err != nil && errors.IsNotFound(err) {
```

Comment:
Nit, spelling and best practice to "pair" the error declaration with the check
```suggestion
	// check if target config exists, create if not
	err := r.Get(ctx, client.ObjectKeyFromObject(target), &corev1.ConfigMap{})
	if err != nil && errors.IsNotFound(err) {
```